lib-ir/src/lib.rs

```diff
@@ -192,10 +193,12 @@ pub enum ExprKind {
         func: Box<Expr>,
         args: Box<[Expr]>,
         location: SourceLocation, // will be displayed in the error message if the signature mismatches
+        is_tail: bool,
     }, // function application (operators are functions too).  Closure parameter is implicitly prepended to the argument list.  Called using Source indirect calling convention (closure, length, and callerid as i32 params; others are Any and on unprotected stack). Static type of func must be func.
     DirectAppl {
         funcidx: FuncIdx,
         args: Box<[Expr]>,
+        is_tail: bool,
```

Comment:
I wonder if being a tail call is different enough to warrant a separate enum variant (e.g. `ReturnAppl` and `DirectReturnAppl`), since a ReturnAppl is effectively a Appl followed by a Return.  For example, `heap_encode_prologue_epilogue` is necessary for Appl but unnecessary for ReturnAppl (we could of course share code in `encode_appl()` and `encode_return_appl()` by calling a common function to do the common things).

Support from other languages:  WebAssembly uses separate op-codes for return_call and return_call_indirect.  Rust has reserved the `become` keyword for tail calls, hence treating it syntactically differently from normal calls.